### ISSUE TITLE
GDscript regression - Typed iteration

### ISSUE THREAD

**Issue by Alvati** [godotengine/godot#39716]
**Godot version:**
Works in 3.2.1 (b0eca5828ccd3a5c07b77e82e567c45937de9056)
Doesn't work since 3.2.2 (5ee9553591ebb7926a238f2d5b5fb154db602b95)

**Issue description:**
This :
```
var test: int
for test in range(3):
	pass
```
... no longer works.

**Old behaviour:**
	If the variable ("test" here) already exists, use it, else create a new one with the scope of the loop.
**New behaviour:**
	Always try to create a new variable, if it already exists, fail with:

> Variable "test" already defined in this scope

Being able to type the iteration variable is important, as Arrays don't support element typing (gdscript arrays are `Array<Variant>`, if you do implement that, please remember Dictionnary.keys()).

As to why type your iteration variable, it enables completion and points out rocks that may have fallen in your array.

--------------------
**Anoying workaround:**
As far as I am aware, now you would need something like this to type your variable :
```
var test: int
for tmp in range(3):
	test = tmp
```
... which is anoying at best, and generally confusing (looking at the loop wouldn't tell you the name of the variable actually used).

--------------------

**Other usecases:**
	If you want to break compatibility and add something like `for test: int in range(3):`, there is still at least one unadressed usecase left :

```
var frog := player
for frog in swamp:
	if frog.matches(criteria):
		break
print(frog)
```
Here, by using the old system, you get a default value if the array is empty and know on what you used break whithout having to introduce another variable.

--------------------
**Conclusion:**

I hope you either restore or reimplement typed iteration. If I have to change every single loop in my project, at least I would like them to still be typed properly.

**Comment by akien-mga** [comment]
CC @vnen @ThakeeNathees

**Comment by ThakeeNathees** [comment]
> Old behaviour:
If the variable ("test" here) already exists, use it, else create a new one with the scope of the loop.

No, the old behavior was if a variable already exists in the outer scope it creates a **new** variable with the same name in the scope, it doesn't use it.

In GDScript a block should not declare a variable with the same name as a variable declared in any block that contains it. for an example:
```
var test: int
while true:
	var test ## <-- this is invalid
```

> var test: int
for test in range(3):
	pass

it's not typed iteration, even if the type of the first definition is different it doesn't throw any error (cause both are different variables with the same name) and typed iteration and typed signal arguments doesn't supported yet (thus the other cases in invalid)

also that't not a workaround it's the only way to do if you want typed iteration.

> I hope you either restore or reimplement typed iteration. If I have to change every single loop in my project, at least I would like them to still be typed properly.

I think you're doing it wrong as there is no such thing as typed iteration, but you could open a proposal for it. typed array world be great too.

**Comment by Alvati** [comment]
Just tried it out, and you're right, they are separate variables
```
var color := Color.aliceblue
for color in [Color.antiquewhite]:
    color = 3
    print(color)
print(color)
```
Prints `3` and `0.94,0.97,1,1`

However, the inner variable still gets the completion as if it was the outer one.

**Comment by aaronfranke** [comment]
This isn't a bug, this is intentional behavior. GDScript should not allow you to shadow previously defined variables like this, and the reported 3.2.2 behavior is exactly that. If your 3.2.1 project breaks in this way, that code should be considered a bug in your project and you should fix it regardless of 3.2.2.
